### DISCUSSION TITLE
How to comment out some contents in .qmd files?

### DISCUSSION THREAD

**Discussion by marcuswang6** [quarto-dev/quarto-cli#3330] [Feature Requests]
If we use `#`, it makes word larger.

**Comment by cscheid** [comment]
The HTML comment syntax is supported by Pandoc and should work in all output formats:

```
<!--
your comment goes here
-->
```

**Reply by allefeld** [reply]
That works for hiding in the output, but code blocks in an HTML comment still get executed.

**Reply by mcanouil** [reply]
See https://github.com/quarto-dev/quarto-cli/discussions/3330#discussioncomment-4150370

**Reply by cderv** [reply]
> but code blocks in an HTML comment still get executed.

@allefeld If you are using **knitr** as a Quarto computation engine, you can leverage the `comment` engine to wrap your whole commented blocks 
https://yihui.org/en/2022/01/knitr-news/#the-new-engines-comment-verbatim-and-embed

It was meant exactly for the purpose of commenting part of a document that includes code chunks that should not be evaluated.

**Reply by allefeld** [reply]
I'm using Python via Jupyter.

**Reply by allefeld** [reply]
After spending an hour reading through different issues, discussions etc., I'm no closer to a solution.

Can we please have a way to comment out parts of a `qmd` file, __including suppression of execution__?

If it helps, I'd be happy to create a fresh issue or discussion based on the following.

### why

In my experience, the ability to 'comment out' is a crucial aspect of programming, especially debugging. One might argue that a `qmd` file is not program code itself but only contains program code, but I think the same applies. A `qmd` file is similar to a code file, just reversed (instead of inserting explanations in code as comments, code is inserted in the explanation as code blocks). When developing more complex computations split over several code blocks, being able to quickly get rid of some of them temporarily makes development much easier.

### previous discussions

There was a feature request https://github.com/quarto-dev/quarto-cli/issues/696, but it was closed by the author after it was pointed out that a div with class `content-hidden` does not appear in the output – but that only applies to visibility, not to execution. It even lead to a PR https://github.com/quarto-dev/quarto-cli/pull/1906 which however ended up being discarded.

There are two suggestions to abuse Pandoc syntax like reference links https://github.com/quarto-dev/quarto-cli/discussions/3330#discussioncomment-4150370 or YAML blocks https://github.com/quarto-dev/quarto-cli/discussions/3330#discussioncomment-4149782 – but those are abuses, there should be a proper solution.

I also found suggestions to use Pandoc's Lua filters https://github.com/quarto-dev/quarto-cli/issues/696#issuecomment-1103935131, but that doesn't help because Pandoc only comes into play after code execution.

### how

A few options for the syntax come to mind:

1) a regular HTML comment `<!-- -->`
2) a 'triple-dash' comment `<!--- --->` as suggested at https://github.com/quarto-dev/quarto-cli/issues/696#issuecomment-1103795774 (inspired by ColdFusion?)
3) a div with class `comment` as in <https://github.com/quarto-dev/quarto-cli/pull/1906>
4) a code block with 'language' `comment` like [in knitr](https://yihui.org/en/2022/01/knitr-news/#the-new-engines-comment-verbatim-and-embed)

To my mind it is less important how it is written, though I would prefer option 1 or 2.

What is however important is that everything marked as a comment either is not included at all in the `ipynb` file which Quarto creates for execution, or is included within a text cell.

**Reply by baptiste** [reply]
I think the underlying problem here is that pandoc parses the AST [after the computational engine has done its work](https://github.com/quarto-dev/quarto-cli/discussions/541), and the code (chunks and inline) are therefore identified (in the form of cells, or via regex) and executed before pandoc has had a chance to flag the content as a comment (and thereby disable execution). 

One possible approach, I think, would be to allow for a [first pass of pandoc with a custom writer](https://github.com/quarto-dev/quarto-cli/discussions/3150#discussioncomment-4044195), almost doing a no-op (i.e. returning almost the same output as the input), but performing a few tasks on the input document that are not easily doable in the current system. This could include detecting a comment _before_ any code is executed, and reformatting everything inside so that it won't get executed (or indeed even seen by the engine). 
Note that this is conceptually different from pre-processing scripts, which afaik are not meant to transform qmd input files.

**Reply by cderv** [reply]
Not meant as an answer but just to add to the context and discussion: 

- Suppression of execution can be done using `eval: false` on chunk. It works for jupyter and knitr. 
It is not a comment character but it will act as commenting. You can apply in batch though. 

- for jupyter, .qmd is converted to .ipynb for execution. And ipynb is a valid `quarto render` input. How would you comment a code cell in Jupyter .ipynb ? Is this possible ? 
Just curious of what exist already. 

- Pre processing .qmd with filtering operation is a recurring topics. It is not as simple because as explained above current AST processing happens after execution. Though this is probably something to keep thinking about if ever possible.

**Reply by mcanouil** [reply]
Conditional content to the rescue?

````qmd
---
title: "Quarto Playground"
format: html
---

This is a playground for Quarto.

:::{.content-hidden}

Whatever you want ...

```{r}
#| eval: false

file.create("hello.world")
```

:::

````


<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/0591feca-17d2-4b4f-9845-57fe58ca70b9">

**Reply by baptiste** [reply]
````
---
title: "Quarto Playground"
format: html
---

This is a playground for Quarto.

:::{.content-hidden}

Whatever you want ...

```{r}
#| eval: false

file.create("hello.world")
```

but beware, `r .NotYetImplemented()` still executed
:::

````

**Reply by allefeld** [reply]
Thanks for considering this.

@baptiste: Yes, that is the problem. I'm not sure how identification of code blocks is currently implemented, but my impression is that it is something like regexp-based? I don't think an initial Pandoc pass is absolutely necessary though, just a first regexp-pass should do: 1) filter out everything from `<!--` to the next `-->` (that doesn't support nested comments, but as far as I know neither do they in HTML) 2) extract the code blocks as before. If an initial Pandoc pass would be that quick though, sure, why not.

@cderv:

- Suppression of execution via `eval: false` is possible, but tedious. In other languages, commenting out a block is something that can be done with a button press, e.g. Ctrl+Shift+7 in VS Code.

- The comment would either be excluded from the ipynb entirely, or it could be put in a text cell.

@mcanouil: Conditional content by itself doesn't help, the code blocks are still executed, as you acknowledged by the explicit `eval: false` in your example.

**Reply by mcanouil** [reply]
To note, in many languages (if not all), commented code is not considered a good practice[^1].

[^1]: I don't say there should not be a way to do this.

**Reply by cderv** [reply]
> I'm not sure how identification of code blocks is currently implemented, but my impression is that it is something like regexp-based?

@allefeld Yes with **knitr** engine it is regex based, for blocks and inlines. That is why **knitr** introduced a way to comment part of documents using 
`````markdown
````{comment}
Anything here will be ignored by **knitr**, so code blocks and inline code will be ignored
````
`````

Quarto relies on **knitr** when `engine: knitr`. 

For Jupyter (when `engine: jupyter`), it works differently based on Jupyter notebook parsing and execution. As I said previously, I don't know if a Jupyter notebook offers a way to comment some part of its content including code cells to avoid there execution. If so, we could probably leverage that. 

More broadly, a way to do .qmd commenting, including avoiding code computation and engine independent, would mean having a way to filter out some content prior to any code computation execution. Pre-engine filter is something that we want to tackle at some point, and maybe when possible, this type of feature would be possible. 

Anyhow, I don't think there is much more to say - we cover quite some ground on current state and existing ones. this is a feature request, and we'll keep that in mind for our next versions. 

For reference, discussion in Pandoc regarding comments in Markdown 
* https://github.com/jgm/pandoc/issues/7762

**Reply by mcanouil** [reply]
The idea of a pre-filter could be helped[^1] with the work that is going to be done in the future, see:
- https://github.com/quarto-dev/quarto-cli/issues/6518

[^1]: blocks comments could break between files dependencies if not used properly by users and Quarto would have almost way to know about this until too late.

PS: I am changing this from Q&A to Feature Request.

**Reply by cscheid** [reply]
I think the best way to do this is to rely on two things:

- we already know we need to add an additional Pandoc pass that happens before engines (so that filters can add code cells that will be executed, and remove code cells as well)
- The "triple-dash html comment" `<!---` instead of `<!--` is a special Pandoc construct that removes the comment from the output. This distinction is important for Pandoc, which needs to handle comments that are intended to appear in the output document ("regular" HTML comments), and comments that denote "remove this from the output document".

So, when we implement the additional pass, we will get, for free, the triple-dash HTML comments.

(Your next question, I'm sure, is "when are we getting the additional pass?" The best estimate I can give you is quarto 1.6)

**Reply by allefeld** [reply]
Ok great, good to know it's on its way!

As you've probably noticed, https://github.com/quarto-dev/quarto/issues/252 is related. I closed it, but it still got some attention.

**Reply by allefeld** [reply]
Just to say, you guys are the most responsive. Greatly appreciated!

**Reply by allefeld** [reply]
Another thought: If the initial Pandoc pass is coming anyway, maybe it would be a good idea to use that not just for preprocessing, but to create the `ipynb`  (via a a custom writer)?

Because another thing I noticed when investigating this is that currently, nesting of code blocks tagged for execution within regular code blocks is not respected. Meaning in
`````````markdown
``````
```{python}
2+2
```
``````
`````````
the inner Python still gets executed
`````````markdown
::: {.cell execution_count=1}
``` {.python .cell-code}
2+2
```

::: {.cell-output .cell-output-display execution_count=112}
```
4
```
:::
:::
`````````
i.e. there is no proper parsing. Using Pandoc to extract the executed blocks would fix that, because the content of the enclosing code block would not appear as a code block itself in the AST.

**Reply by cscheid** [reply]
Yes, that's already part of the plan (and is part of why this is a bigger change that can't just be done right away)

**Reply by cderv** [reply]
@allefeld for this, there is a special syntax to apply on the block
https://quarto.org/docs/computations/execution-options.html#unexecuted-blocks

```````
``````
```{{python}}
2+2
```
``````
```````

This is because 
````
```{python}
````

means executable code block

**Reply by cscheid** [reply]
@cderv But @allefeld does have a very good point here, which is that the inner python block is inside an outer code block. This happens only because knitr incorrectly uses regular expressions everywhere. It's way too late to change this behavior in knitr, but we should absolutely be using a real parser everywhere...

**Reply by cderv** [reply]
Oh yeah for sure. I just wanted to explain what currently works, and which feature we added (the `{{engine}}` syntax) to allow that. 

knitr has made a choice to not consider Markdown structure (and same for other language it supports like Rnw or Rhtml) and R Markdown kept that by considering Markdown structure after knitting. This allows to create some content in document dynamically while mixing markdown verbatim and computation code.

This assumption can definitely by reversed from Quarto and it would made much sense. Computation code would then need to express all the markdown content verbatim. 

Definitely agree that we need a parser. And maybe not too late for **knitr** as we are thinking of a Rmd parser too 😉

Really great discussion @allefeld ! thanks for feedback and raising all those points!

**Reply by baptiste** [reply]
There's two things in the proposal, I think: one is to have pandoc parse the input as an AST before computations, the other is to have two passes of pandoc. I believe both bring extra flexibility, and the combination would be very versatile indeed.

It would be possible, I guess, to simply reverse the order: have pandoc create an AST first, then the engine processes code etc. and injects the results into the AST (with new leaves, branches, etc. or even modifying existing branches). Between the engine and Lua filters any kind of transformation can be done on the AST, technically speaking. But in practice there are situations where it's also useful, to simplify the logic and implementation, to have two separate parsing steps occurring sequentially, with computations in-between.

**Reply by cscheid** [reply]
That means a completely different system than quarto is today; pretty much all of our filters operate on engine outputs, so this is a non-starter.

**Reply by baptiste** [reply]
Indeed, thankfully with a two-pass approach it becomes possible for Alice to share the cake and cut it later too. 

<img width="734" alt="Screenshot 2023-08-26 at 11 02 15" src="https://github.com/quarto-dev/quarto-cli/assets/18970/72e989a0-9c2e-40a3-8d79-2c91e1ac63e7">

**Comment by baptiste** [comment]
A lesser-known trick is that you can have Yaml blocks within the document, and comment out everything inside with `#`

````

---
# When you click the **Render** button a document will be generated that includes both content and # the output of embedded code. You can embed code like this:
# 
# ```{r}
# 1 + 1
# ```
# but beware, `r .NotYetImplemented()` still executed
---

````

Though, unfortunately, inline R code `r ...` still gets executed as knitr grabs those via regex, and is not aware of the commented context (same applies to the official html commenting construct).

**Reply by mcanouil** [reply]
`---` separated with blank lines is horizontal rules for `html`, it's "new slide" in `revealjs`.
If `---` underlines text it can also be a level 2 header.

Why would someone add a new layer of confusion and error by using YAML block syntax to make comment blocks.
I would personally never recommend this to anyone.

**Reply by baptiste** [reply]
I believe it's legit in pandoc (that's where I first encountered the suggestion), and has the advantage of using consistent `# ` markers for comments, regardless of the content (code, or text). I'm not saying this is a good solution, but it's a possible option if someone prefers this style of commenting to `<!--`. I don't see pandoc confusing it for titles or slide breaks, which have specific empty line requirements I believe.

**Reply by mcanouil** [reply]
I know you are not saying it's a good solution. 😉
I definitely find the HTML style comments more versatile, since it only requires to open and close the block or line without the need to add extra characters.
I am curious to know the rational for Pandoc to implement/allow the syntax you described.
I did not find anything on this syntax.

Regarding the inline R code, I believe the issue is the same with HTML (block) comments.

**Reply by baptiste** [reply]
IIRC pandoc simply consolidates all Yaml blocks into one, wherever they might appear in the document, and since `# ` is the Yaml comment character, those lines don't produce anything.

**Reply by mcanouil** [reply]
Oh, that's more of a trick than an actual comments block.
Thanks for the explanation.

**Reply by cderv** [reply]
> Though, unfortunately, inline R code r ... still gets executed as knitr grabs those via regex, and is not aware of the commented context

FWIW Knitr introduced a syntax for commenting blocks: https://github.com/quarto-dev/quarto-cli/discussions/3330#discussioncomment-6814888

**Comment by baptiste** [comment]
(By the way, [djot](https://djot.net/syntax.html#comment) introduces comments that are not tied to HTML syntax – hopefully this will eventually make its way into pandoc as the preferred input format)

**Reply by cscheid** [reply]
This is very interesting indeed.

**Comment by mcanouil** [comment]
There is also a less-known comments trick[^1] which contrary to HTML style won't be displayed in the final document.

```md
[]: # (This is a comment)
[]: # "And this is a comment"
[]: # 'Also this is a comment'
[//]: # (Yet another comment)
[comment]: # (Still another comment)
```

[^1]: Since the trick uses reference link, it will possibly raise some warnings, especially for duplication if several in the documents, *i.e.*, `[WARNING] Duplicate link reference '[]'`.

**Reply by rodrigo-j-goncalves** [reply]
For revealjs presentations: 

`[]: this is a comment.`
`[]:this is also a comment.`
`[some text here]: Yep, also a comment.`
and
`[some text here]: (This is` 
`a multiline `
`comment)`

![image](https://github.com/user-attachments/assets/add38940-d336-4ba0-a25c-e9c55a8c99e2)

**Comment by allefeld** [comment]
@cscheid:
> * The "triple-dash html comment" `<!---` instead of `<!--` is a special Pandoc construct that removes the comment from the output. This distinction is important for Pandoc, which needs to handle comments that are intended to appear in the output document ("regular" HTML comments), and comments that denote "remove this from the output document".
>
> So, when we implement the additional pass, we will get, for free, the triple-dash HTML comments.

I couldn't find that in the Pandoc documentation. Apparently at one point a special comment syntax was discussed, (see https://github.com/jgm/pandoc/issues/2552), but in the end only the [`--strip-comments` option](https://pandoc.org/MANUAL.html#option--strip-comments[) was implemented, which filters out regular HTML comments.

**Reply by cscheid** [reply]
Thanks for following up! You're of course completely right. The `<!---` comment was definitely a misconception that I (and other folks here) shared; the fact that it doesn't exist means that we should certainly support _some_ comment option. Unfortunately, to properly support comments we'd need an actual parser used consistently throughout Quarto, and that doesn't yet exist.

**Comment by sijow** [comment]
Just a note that @cderv [proposed](https://github.com/quarto-dev/quarto-cli/issues/696#issuecomment-1256636144) a solution that works well for me: using `{.content-hidden}` blocks.

**Comment by floholl** [comment]
Sorry for bumping an old thread, but since this seems to be an ongoing discussion (to at least some degree), maybe a glance over to [org-mode](https://orgmode.org/) can meaningfully inform future design decisions.

I use and love both, Quarto and org-mode, but I think of the latter as the more flexible tool when it comes to the *pre*-publication stage of a document, where I am still *developing* my ideas (as opposed to preparing them for publication).

Why is this relevant for this discussion? Because I can boil it down to three specific org-mode features:

- The ability to easily comment individual lines with `#`.
- But also: The ability to [exclude from export any sections whose headings are marked as TODO tasks](https://emacs.stackexchange.com/a/18838).
- The ability to conveniently re-order (sub)sections with *Alt+↑* and *Alt+↓* (arguably an editor feature).

Fair enough, Quarto is a publishing system, and not a task manager in the way that org-mode is. But boy, would I party if  Quarto was to adopt `.org` as a supported *input* format in addition to `.qmd` and `.ipynb`. :smile_cat:

**Reply by cscheid** [reply]
We're unlikely to ever do this - it's a huge amount of added complexity for, unfortunately, a very small audience. With that said, you could use project pre-render scripts and Pandoc to convert from .org to .md and then to .qmd.
